### PR TITLE
Update to handle multiple datatypes

### PR DESCRIPTION
**Pull request by jwittenbach**
WIP

### CODE REVIEW

File: factorization/base.py [hunk 1]
```diff
@@ -5,27 +5,38 @@ class Algorithm(object):
...
+        elif isinstance(X, (BoltArraySpark, ndarray)):
+            if X.ndim != 2:
+                raise ValueError("Array to factor must be 2-dimensional")
             data = X
-        else:
-            data = fromarray(X)
 
-        if data.mode == "local":
-            results = self._fit_local(data)
+        # Factor
```
Comment (freeman-lab) [inline factorization/base.py]: minor nit but we usually us all lowercase for comments

File: test/test_algorithms.py [hunk 2]
```diff
@@ -65,30 +65,30 @@ def test_svd(eng):
     x1 = fromarray(x)
     x2 = fromarray(x, engine=eng)
 
-    u1, s1, v1 = to_array(SVD(k=2, seed=0).fit(x1))
-    u2, s2, v2 = to_array(SVD(k=2, seed=0, method="direct").fit(x2))
+    u1, s1, vT1 = to_array(SVD(k=2, seed=0).fit(x1))
```
Comment (freeman-lab) [inline test/test_algorithms.py]: i might use `v1_T` instead

File: test/test_algorithms.py [hunk 3]
```diff
@@ -65,30 +65,30 @@ def test_svd(eng):
...
+    u2, s2, vT2 = to_array(SVD(k=2, seed=0, method="direct").fit(x2))
 
     tol = 1e-2
     assert allclose_sign(u1, u2, atol=tol)
     assert allclose(s1, s2, atol=tol)
-    assert allclose_sign(v1, v2, atol=tol)
+    assert allclose_sign(vT1.T, vT2.T, atol=tol)
```
Comment (freeman-lab) [inline test/test_algorithms.py]: you're transposing and transposing back? what's going on here?

File: factorization/algorithms/ICA.py [hunk 4]
```diff
@@ -43,12 +45,12 @@ def _fit_spark(self, data):
                             " must be less than the data dimensionality " + str(ncols))
 
         # reduce dimensionality
-        u, s, v = SVD(k=self.k_pca, method=self.svd_method, seed=self.seed).fit(data)
-        s, v = s.toarray(), v.toarray()
+        u, s, vT = SVD(k=self.k_pca, method=self.svd_method, seed=self.seed).fit(data)
```
Comment (freeman-lab) [inline factorization/algorithms/ICA.py]: slight preference `v_T` for transpose

**Comment by freeman-lab** [comment]
@jwittenbach Is this basically good to go? Want me to look anything over?

**Comment by jwittenbach** [comment]
@freeman-lab just need to add some tests for the various reshaping function and then add some documentation for it. Working on it right now actually :smile_cat:

**Comment by freeman-lab** [comment]
@jwittenbach nice!

**Comment by jwittenbach** [comment]
@freeman-lab alright, I think this is getting really close. If you wouldn't mind taking a look, that would be great! The tests have been pretty seriously reworked and reflect the new API, so that might be a good place to start.

**Comment by freeman-lab** [comment]
After an offline chat and further thoughts, I think this is all great, only suggestion is that we make all output types consistent, specifically, I think they should all be made local numpy arrays by default, shaped using the same rules as now. But if a flag is set, maybe `return_parallel` or something, they should all come back as distributed bolt arrays. That just seems consistent and easier to explain and maintain than the current output rules, which seem somewhat tailored to the current algorithms.

**Comment by jwittenbach** [comment]
@freeman-lab implemented the changes we talked about. All return values are now either `ndarray` or `BoltArray`. The only time you get the `BoltArray` is on the applicable arrays when you set the optional kwarg `return_parallel=True` (default is `return_parallel=False`).

**Comment by freeman-lab** [comment]
@jwittenbach great! One last thing: you mentioned there was a separate set of tests for the various shaping operations, but I don't see that in the PR, maybe you just forgot to add the file?

**Comment by jwittenbach** [comment]
@freeman-lab hah, good call -- that's exactly what happened. Added.

**Comment by freeman-lab** [comment]
great! let's merge this in! 🚢

**Pull request merged**